kinetics_MO.py

```diff
@@ -0,0 +1,261 @@
+
+# use MeasurementOptimizer to pre-compute the unit FIMs
+calculator = MeasurementOptimizer(
+    jac_info,  # SensitivityData object
+    measure_info,  # MeasurementData object
+    error_cov=error_cov,  # error covariance matrix
+    error_opt=CovarianceStructure.measure_correlation,  # error covariance options
+    print_level=3,  # I use highest here to see more information
+)
+
+# calculate a list of unit FIMs
+calculator.assemble_unit_fims()
+
+
+### MO optimization framework
+
+# optimization options
+mip_option = True
+objective = ObjectiveLib.D
+fixed_nlp_opt = False
+mix_obj_option = False
+alpha_opt = 0.9
+
+sparse_opt = True
+fix_opt = False
+small_element = 0.0001  # the small element added to the diagonal of FIM
+file_store_name = "test_run_"
+
+num_dynamic_time = np.linspace(0, 60, 9)
+
+static_dynamic = [[0, 3], [1, 4], [2, 5]]  # These pairs cannot be chosen simultaneously
+time_interval_for_all = True
+
+# map the timepoint index to its real time
+dynamic_time_dict = {}
+for i, tim in enumerate(num_dynamic_time[1:]):
+    dynamic_time_dict[i] = np.round(tim, decimals=2)
+
+# give range of budgets for this case
+budget_ranges = np.linspace(1000, 5000, 11)
+# give a trial ranges for a test; we use the first 3 budgets in budget_ranges
+trial_budget_ranges = budget_ranges[:3]
+# initialize with A-opt. MILP solutions
+# choose what solutions to initialize from:
+# minlp_D: initialize with minlp_D solutions
+# milp_A: initialize with milp_A solutions
+# lp_A: iniitalize with lp_A solution
+# nlp_D: initialize with nlp_D solution
+initializer_option = "lp_A"
+
+# ==== initialization strategy ====
+# according to the initializer option, we provide different sets of initialization files
+if initializer_option == "milp_A":
+    # all budgets
+    curr_results = np.linspace(1000, 5000, 11)
+    # initial solution file path and name
+    file_name_pre, file_name_end = "./kinetics_results/MILP_", "_a"
+
+elif initializer_option == "minlp_D":
+    # all budgets
+    curr_results = np.linspace(1000, 5000, 11)
+    # initial solution file path and name
+    file_name_pre, file_name_end = "./kinetics_results/MINLP_", "_d_mip"
+
+elif initializer_option == "lp_A":
+    # all budgets
+    curr_results = np.linspace(1000, 5000, 41)
+    # initial solution file path and name
+    file_name_pre, file_name_end = "./kinetics_results/LP_", "_a"
+
+elif initializer_option == "nlp_D":
+    # all budgets
+    curr_results = np.linspace(1000, 5000, 41)
+    # initial solution file path and name
+    file_name_pre, file_name_end = "./kinetics_results/NLP_", "_d"
+
+# initialize the initial solution dict. key: budget. value: initial solution file name
+# this initialization dictionary provides a more organized input format for initialization
+initial_solution = {}
+# loop over budget
+for b in curr_results:
+    initial_solution[b] = file_name_pre + str(int(b)) + file_name_end
+
+# ===== run a test for a few budgets =====
```

Comment:
This is duplicate code. Instead, you should setup the input data above and then run the analysis. Make it easy for the user to toggle between running just a few cases and reproducing all of the paper results.

Reply:
`rerun_all_paper_results=True`
if rerun_all_paper_results: # define input data 
else: # define a sensitivity budgets to test

Reply:
done